묵찌빠게임 [STEP2] Hisop, 쥬봉이

jyubong opened the pull request:
@hojun-jo 
안녕하세요 Step2 PR입니다. 잘부탁드립니다 :)

---
### 고민이 되었던 점
- **첫번째 게임의 가위(1), 바위(2) 순서에서 두번째 게임에서는 묵(1), 찌(2)로 1, 2번의 순서가 바뀌었습니다. 이를 위한 여러 방안을 생각해보았으며 그 중 아래 방법으로 택했습니다.** 아래의 방법이 유사한 기능을 가진 함수를 묶어 기능분리에도 좋은 방법이라고 생각해서 택하였습니다.
    - `컴퓨터와 유저선택 비교 함수` : 가위바위보와 묵찌빠에서 비교를 구현하는 로직 이 비슷
    - `유저의 input값을 첫번째 게임에서는 1을 가위, 2를 바위로, 두번째 게임에서는 1을 바위, 2를 가위로 변경해 주는 함수`

- **기능에 따라 함수가 잘 분리되었는지 고민했습니다.**
   - 게임 실행 / 유저입력 매핑 / 승패 구분 / 결과 출력으로 분리했습니다.
- **세세하게 나눠지는 분기의 처리를 고민했습니다**
   - 유저입력의 매핑은 switch로 처리했습니다.  
   - 함수를 호출할 때 상황에 따라 다른 매개변수를 전달해야하는 '현재 턴'은 input으로 변수의 주소를 전달하도록 했습니다.  
   (if 분기로 함수의 호출을 나눴을 때 전달하는 값이 직관적이지 않다고 생각해 특정 상황에 '현재 턴'을 변경하고 항상 턴을 전달하도록 했습니다) 


### 조언을 얻고 싶은 부분
- 언제나 네이밍이 고민됩니다.
- 현재 라운드에 따른 결과출력함수를 다른 이름으로 작성했는데 오버로딩으로 작성해보는건 어떻게 생각하시나요?

Commit:
feat: enum추가, 결과 출력 함수 추가

RoundResult, Choice enum추가
라운드별 결과를 출력하는 함수 추가

jyubong committed:
refactor: 승자 결정 함수 수정

compareWithComputer -> deciveVictory
computerChoice 랜덤값 생성: Int.random -> allCases.randomElement
computer와 user choice 비교 수정 : String -> enum

Commit:
feat: mappingUserChoice 추가

userInput을 라운드에 따른 enum case로 mapping하는 함수 추가

jyubong committed:
refactor: startGame -> playRoundOne 수정

Commit:
feat: playRoundTwo 함수 생성

jyubong committed:
refactor: enum skip -> startGame, displayRoundTwo default 수정

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,48 +4,149 @@
+    case draw
+    case invalidInput
+    case gameStart
+}
```
invalidInput과 gameStart라는 케이스가 게임 결과라는 타입에 적절한 케이스일까요?
 게임 진행(분기)의 판단을 위한 타입이라면 다른 이름을 생각해 보시는 것도 좋을 것같습니다!

jyubong replied:
게임 진행 판단의 의미로 RoundResult -> GameProgress 로 변경하였습니다.

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,48 +4,149 @@
+    case rock
+    case paper
+    case scissors
+}
```
가위, 바위, 보를 직관적으로 나타낼 수 있는 이름이 좋을 것같습니다.

jyubong replied:
Choice -> RockPapaperScissors로 변경하여 가위,바위,보를 나타내었습니다.

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,48 +4,149 @@
-    guard let userChoice = readLine() else {
-        print("잘못된 입력입니다. 다시 시도해주세요.")
-        return startGame(isContinous: true)
+func playRoundOne(result: RoundResult) {
```
위의 RoundResult 코멘트와 연결 지을 수 있을 것 같습니다. 게임을 시작하는데 결과가 필요한 건 어색해 보이네요.
 그리고 RoundOne 대신에 가위바위보라는 것을 알 수 있는 이름을 사용하는 것이 좋을 것 같습니다

Hi-sop replied:
playRockPaperScissorsGame으로 변경했습니다.

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,48 +4,149 @@
+    }
+    
+    playRoundTwo(result: decideVictory(userChoice: userChoice), turn: &turn)
+}
```
inout까지 활용하셨군요! 이 부분은 원하시는 의도대로 잘 구현하신 것같습니다. inout 활용을 위해 공부하신 내용을 알려주실 수 있을까요?
 
 제 생각에는 turn 파라미터 없이 이전 게임의 결과만으로 조금 더 단순하게 만들 수 있을 것같습니다. 그러면 if문도 지울 수 있을 것같네요.
 한 번 시도 해보시는 것도 좋을 것같습니다!
 
 그리고 RoundTwo 대신에 묵찌빠라는 것을 알 수 있는 이름을 사용하는 것이 좋을 것같습니다

Hi-sop replied:
일반적으로 함수는 파라미터로 상수를 복사해서 받습니다. inout을 사용하지 않으면 주소가 아닌 값만 복사하게 됩니다. inout은 해당 파라미터가 주소(참조)를 전달한다고 명시하는 역할을 합니다.
깊은 복사 개념과 마찬가지로 전달받은 함수에서 값을 변경하는 것은 전달한 원본 값에도 영향을 미칩니다. 

https://docs.swift.org/swift-book/documentation/the-swift-programming-language/functions/#In-Out-Parameters

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,48 +4,149 @@
+        default:
+            return nil
+        }
     }
```
👍

hojun-jo reviewed (changes requested):
히섭, 쥬봉이 안녕하세요~
스탭2 수고하셨습니다!
이번 스탭도 잘 하고 계신 것같아 크게 수정할 부분은 별로 없을 것같습니다!

---
## 칭찬 드리고 싶은 점
- 커밋 단위를 잘 나누고 계신 것 같네요!👍
- 컨벤션도 계속 잘 맞추고 계시네요👍
- 기능 분리도 잘 되어 보입니다👍

---
### 고민이 되었던 점

> 첫번째 게임의 가위(1), 바위(2) 순서에서 두번째 게임에서는 묵(1), 찌(2)로 1, 2번의 순서가 바뀌었습니다. 이를 위한 여러 방안을 생각해보았으며 그 중 아래 방법으로 택했습니다. 아래의 방법이 유사한 기능을 가진 함수를 묶어 기능분리에도 좋은 방법이라고 생각해서 택하였습니다.

> 컴퓨터와 유저선택 비교 함수 : 가위바위보와 묵찌빠에서 비교를 구현하는 로직 이 비슷
유저의 input값을 첫번째 게임에서는 1을 가위, 2를 바위로, 두번째 게임에서는 1을 바위, 2를 가위로 변경해 주는 함수
기능에 따라 함수가 잘 분리되었는지 고민했습니다.

> 게임 실행 / 유저입력 매핑 / 승패 구분 / 결과 출력으로 분리했습니다.
세세하게 나눠지는 분기의 처리를 고민했습니다

- 함수 기능 분리 잘 하고 계십니다👍

> 유저입력의 매핑은 switch로 처리했습니다.
함수를 호출할 때 상황에 따라 다른 매개변수를 전달해야하는 '현재 턴'은 input으로 변수의 주소를 전달하도록 했습니다.
(if 분기로 함수의 호출을 나눴을 때 전달하는 값이 직관적이지 않다고 생각해 특정 상황에 '현재 턴'을 변경하고 항상 턴을 전달하도록 했습니다)

- 이 부분은 원하시는 의도대로 잘 구현하신 것같습니다!

### 조언을 얻고 싶은 부분

> 언제나 네이밍이 고민됩니다.
- 정답이 없는 부분이라 항상 고민해야 하는 것같습니다. 😅
 API Design Guidelines을 몇 번 읽는 것도 좋습니다!
 함수, 변수의 기능과 역할을 잘 생각해본 후에 먼저 한국어로 여러 후보를 만들고 영어로 바꾸는 것도 괜찮은 방법일 것같습니다.
 지금 잘 하고 계신 부분도 있으니 전달 인자 레이블을 이용해 함수와 인자까지 한 문장으로 읽을 수 있도록 시도 해보시는 것도 좋을 것같습니다!

> 현재 라운드에 따른 결과출력함수를 다른 이름으로 작성했는데 오버로딩으로 작성해보는건 어떻게 생각하시나요?
- 지금처럼 매개변수가 다를 때에는 오버로딩도 괜찮은 방법일 것같습니다.
 아래 코멘트처럼 수정할 수도 있으니 이 부분은 고려 해보시기 바랍니다!

jyubong committed:
refatctor: enum, 함수 이름 수정, playResult함수 오버로딩으로 변경

jyubong committed:
docs: Update README.md

jyubong committed:
docs: Update README.md

Commit:
refactor: playgame함수의 로직을 변경

enum에 printOption을 추가하여 현재 출력 옵션을 전달할 수 있게 함
출력함수 호출에 대한 분기를 명확하게 변경
출력함수 네이밍 변경
Winner enum case에 none을 추가하여 승자가 없는 상태도 전달할 수 있게 함
